source/byte_buf.c

```diff
@@ -242,32 +259,102 @@ static const uint8_t s_tolower_table[256] = {
     220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241,
     242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255};
 
-bool aws_byte_cursor_eq_case_insensitive(const struct aws_byte_cursor *a, const struct aws_byte_cursor *b) {
-    if (!a || !b) {
-        return a == b;
+bool aws_array_eq_ignore_case(const void *array_a, size_t len_a, const void *array_b, size_t len_b) {
+    if (len_a != len_b) {
+        return false;
     }
 
-    if (a->len != b->len) {
-        return false;
+    /* If either is null: true if both are null or the size is 0 */
+    if (!array_a || !array_b) {
+        return (array_a == array_b) || (len_a == 0);
     }
 
-    for (size_t i = 0; i < a->len; ++i) {
-        if (s_tolower_table[a->ptr[i]] != s_tolower_table[b->ptr[i]]) {
+    const uint8_t *bytes_a = array_a;
+    const uint8_t *bytes_b = array_b;
+    for (size_t i = 0; i < len_a; ++i) {
+        if (s_tolower_table[bytes_a[i]] != s_tolower_table[bytes_b[i]]) {
             return false;
         }
     }
 
     return true;
 }
 
-uint64_t aws_hash_byte_cursor_ptr_case_insensitive(const void *item) {
+bool aws_array_eq(const void *array_a, size_t len_a, const void *array_b, size_t len_b) {
+    if (len_a != len_b) {
+        return false;
+    }
+
+    /* If either is null: true if both are null or the length is 0 */
+    if (!array_a || !array_b) {
```

Comment:
if (array_a == array_b) is always equivalent, null or not.

Reply:
In fact, you could do this (perhaps more simply) as:
```c
if (array_a == array_b && len_a == len_b)
```

Reply:
The way it's written, it takes this branch if either is NULL, so there's no chance of dereferencing a nullptr later

the way you wrote it, if one was NULL, it would bypass this check and dereference the nullptr when it entered the for-loop later

Reply:
🙅‍♂️